helpers/helpers.go

```diff
@@ -110,6 +111,183 @@ var ProtoHelpersFuncMap = template.FuncMap{
 	"urlHasVarsFromMessage":   urlHasVarsFromMessage,
 	"lowerGoNormalize":        lowerGoNormalize,
 	"goNormalize":             goNormalize,
+	"leadingComment":          leadingComment,
+	"trailingComment":         trailingComment,
+	"leadingDetachedComments": leadingDetachedComments,
+	"stringFieldExtension":    stringFieldExtension,
+	"boolFieldExtension":      boolFieldExtension,
+	"isFieldMap":              isFieldMap,
+	"fieldMapKeyType":         fieldMapKeyType,
+	"fieldMapValueType":       fieldMapValueType,
+	"replaceDict":             replaceDict,
+}
+
+var pathMap map[interface{}]*descriptor.SourceCodeInfo_Location
+
+func InitPathMap(file *descriptor.FileDescriptorProto) {
+	pathMap = make(map[interface{}]*descriptor.SourceCodeInfo_Location)
+	addToPathMap(file.GetSourceCodeInfo(), file, []int32{})
+}
+
+func InitPathMaps(files []*descriptor.FileDescriptorProto) {
+	pathMap = make(map[interface{}]*descriptor.SourceCodeInfo_Location)
+	for _, file := range files {
+		addToPathMap(file.GetSourceCodeInfo(), file, []int32{})
+	}
+}
+
+func addToPathMap(info *descriptor.SourceCodeInfo, i interface{}, path []int32) {
```

Comment:
Could you maybe add a comment how this recursion ends? I don't quite see it.

Reply:
The AST is a tree as opposed to a graph, so it just traverses the tree until there are no additional nodes to follow.